tensorflow/training/docker/2.14/py3/Dockerfile.cpu

```diff
@@ -0,0 +1,310 @@
+   && ${HOME_DIR}/oss_compliance/generate_oss_compliance.sh ${HOME_DIR} ${PYTHON} \
+   && rm -rf ${HOME_DIR}/oss_compliance*
+
+# remove tmp files
+RUN rm -rf /tmp/*
+
+CMD ["/bin/bash"]
+
+#################################################################
+#  ____                   __  __       _
+# / ___|  __ _  __ _  ___|  \/  | __ _| | _____ _ __
+# \___ \ / _` |/ _` |/ _ \ |\/| |/ _` | |/ / _ \ '__|
+#  ___) | (_| | (_| |  __/ |  | | (_| |   <  __/ |
+# |____/ \__,_|\__, |\___|_|  |_|\__,_|_|\_\___|_|
+#              |___/
+#  ___                              ____           _
+# |_ _|_ __ ___   __ _  __ _  ___  |  _ \ ___  ___(_)_ __   ___
+#  | || '_ ` _ \ / _` |/ _` |/ _ \ | |_) / _ \/ __| | '_ \ / _ \
+#  | || | | | | | (_| | (_| |  __/ |  _ <  __/ (__| | |_) |  __/
+# |___|_| |_| |_|\__,_|\__, |\___| |_| \_\___|\___|_| .__/ \___|
+#                      |___/                        |_|
+#################################################################
+
+FROM common AS sagemaker
+
+LABEL maintainer="Amazon AI"
+LABEL dlc_major_version="1"
+
+ARG TF_URL=https://framework-binaries.s3.us-west-2.amazonaws.com/tensorflow/r2.14_aws/cpu/2024-01-23-01-25/tensorflow_cpu-2.14.1-cp310-cp310-linux_x86_64.whl
+
+# sagemaker-specific environment variable
+ENV SAGEMAKER_TRAINING_MODULE sagemaker_tensorflow_container.training:main
+
+# dependencies for opencv
+# these dependencies are not needed for gpu image
+RUN apt-get update  \
+ && apt-get install -y --no-install-recommends \
+    libgtk2.0-dev \
+ && apt-get install -y -qq libkrb5-dev \
+ && apt-get install -y -qq libsasl2-dev libsasl2-modules \
+ && apt-get install -y -qq krb5-user \
+ && rm -rf /var/lib/apt/lists/* \
+ && apt-get clean
+
+# Need to install wheel before we can fix the pyyaml issue below
+RUN pip install wheel
+
+# https://github.com/yaml/pyyaml/issues/601
+# PyYaml less than 6.0.1 failes to build with cython v3 and above.
+# tf-models-official uses older versions, breaking the install.
+# going to install the older pyyaml and cython to get tfd-models-official
+# the sagemaker package will revert pyyaml back to 6 for its requirement
+# and this is fine since sagemaker is more important than the models and
+# the models still work on pyyaml 6 in this context.
+RUN pip install "cython<3" "pyyaml<6" --no-build-isolation
+
+# https://github.com/tensorflow/models/issues/9267
+# tf-models does not respect existing installations of TF and always installs open source TF
+RUN ${PIP} install --no-cache-dir -U \
+    tf-models-official==2.14.2 \
+    tensorflow-text==2.14.0 \
+ && ${PIP} uninstall -y tensorflow tensorflow-gpu \
+ && ${PIP} install --no-cache-dir -U \
+    ${TF_URL} \
+    "tensorflow-io>=0.33.0,<0.34.0" \
+    tensorflow-datasets
+
+RUN $PYTHON -m pip install --no-cache-dir -U \
+    numba \
+    bokeh \
+    imageio \
+    opencv-python \
+    plotly \
+    seaborn \
+    shap
+
+RUN $PYTHON -m pip install --no-cache-dir -U \
```

Comment:
can we merge this RUN statement with the statement above?

Reply:
In terms of overall image size, there is not much to be gained by merging the two statements.

In terms of dev experience, keeping these two steps separate allows us to plug-and-play the above RUN statement into any other dockerfile. It also keeps two logically different requirements of the dockerfile in two steps.

In terms of image layer management, having massive layers slows down download and extraction time during pulls.

I'll try to address the pros/cons of merging these in a retrospective, but I'd rather keep them separate for now, since it seems to work without issues.